lib/core/src/Cardano/Pool/Ranking.hs

```diff
@@ -0,0 +1,206 @@
+
+-- | Total rewards for a pool if it were saturated.
+--
+-- When a0 = 0 this reduces to just p*R*z0 (tested
+-- by @prop_saturatedPoolRewardsReduces@)
+saturatedPoolRewards :: EpochConstants -> Pool -> Double
+saturatedPoolRewards constants pool =
+    let
+        a0 = getNonNegative $ leaderStakeInfluence constants
+        z0 = unRelativeStake $ saturatedPoolSize constants
+        s = unRelativeStake $ leaderStake pool
+        _R = getNonNegative $ getLovelace $ totalRewards constants
+        p = getNonNegative $ recentAvgPerformance pool
+        -- ^ technically \hat{p} in the spec
+    in
+        (p * _R) / (1 + a0)
+        * (z0 + ((min s z0) * a0))
+
+-- | Determines z0, i.e 1 / k
+saturatedPoolSize :: EpochConstants -> RelativeStakeOf "pool"
+saturatedPoolSize constants =
+    RelativeStake $ 1 / fromIntegral (getPositive $ optimalNumberOfPools constants)
+
+--------------------------------------------------------------------------------
+-- Types
+--------------------------------------------------------------------------------
+
+data EpochConstants = EpochConstants
+    { leaderStakeInfluence :: NonNegative Double
+      -- ^ a_0
+    , optimalNumberOfPools :: Positive Int
+      -- ^ k
+    , totalRewards :: Lovelace
+      -- ^ Total rewards in an epoch. "R" in the spec.
+    } deriving (Show, Eq, Generic)
+
+data Pool = Pool
+    { leaderStake :: RelativeStakeOf "pool leader"
+    , cost :: Lovelace
+    , margin :: Margin
+    , recentAvgPerformance :: NonNegative Double
+      -- ^ An already averaged performance-value
+    } deriving (Show, Eq, Generic)
+
+newtype Lovelace = Lovelace { getLovelace :: NonNegative Double }
+    deriving (Show, Eq)
+    deriving newtype (Ord, Num)
+
+newtype Margin = Margin { getMargin :: Double }
+    deriving (Show, Eq)
+    deriving newtype Ord
+
+unsafeMkMargin :: Double -> Margin
+unsafeMkMargin x
+    | x >= 0 && x <= 1  = Margin x
+    | otherwise         = error $ "unsafeMkMargin: " ++ show x
+                          ++ "not in range [0, 1]"
+
+-- | Stake relative to the total active stake in an epoch
+--
+-- The value
+-- 0.01 :: RelativeStakeOf "pool"
+-- would mean that a pool has a stake that is 1% of the total active stake in
+-- the epoch.
+newtype RelativeStakeOf (tag :: Symbol)
+    = RelativeStake { unRelativeStake :: Double }
+    deriving (Eq, Ord, Show, Generic)
+    deriving newtype (Num, Fractional)
+
+mkRelativeStake :: Lovelace -> EpochConstants -> RelativeStakeOf (tag :: Symbol)
+mkRelativeStake (Lovelace (NonNegative stake)) constants =
+    RelativeStake $ stake / total
+  where
+    total = getNonNegative $ getLovelace $ totalRewards constants
+
+newtype Positive a = Positive { getPositive :: a }
+    deriving (Generic, Eq, Show)
+    deriving newtype (Ord, Num)
+
+unsafeToPositive :: (Ord a, Show a, Num a) => a -> (Positive a)
+unsafeToPositive x
+    | x >= 0    = Positive x
+    | otherwise = error $ "unsafeToPositive: " ++ show x ++ " > 0 does not hold"
```

Comment:
Note that `Positive a` in this module is only used for `Positive Int` which can be substituted with `Natural` from the `base` package. The `Positive` data type is therefore redundant.

Reply:
Sorry, I got the condition wrong here. It should be `x > 0`, which is not captured by `Natural`.

This is to prevent division-by-zero if anyone sets `k = 0`.

I should add tests for these `unsafeTo`-constructors if we intend to have them.